File: shared/python/response_metrics.py
"""Reusable step-response metrics for laboratory data.

The implementation intentionally avoids requiring control-system-specific packages.
It accepts array-like time, process value, and either scalar or array-like setpoint.
"""
from __future__ import annotations

from dataclasses import dataclass, asdict
from typing import Iterable
import math
import numpy as np


@dataclass
class StepMetrics:
    initial_value: float = math.nan
    final_setpoint: float = math.nan
    final_value: float = math.nan
    delay_time_s: float = math.nan
    rise_time_s: float = math.nan
    peak_time_s: float = math.nan
    settling_time_s: float = math.nan
    maximum_overshoot_pct: float = math.nan
    steady_state_error: float = math.nan

    def as_dict(self) -> dict[str, float]:
        return asdict(self)


def _first_crossing(t: np.ndarray, y: np.ndarray, level: float, direction: float) -> float:
    if direction >= 0:
        idx = np.flatnonzero(y >= level)
    else:
        idx = np.flatnonzero(y <= level)
    return float(t[idx[0]]) if idx.size else math.nan


def step_metrics(
    time_s: Iterable[float],
    process_value: Iterable[float],
    setpoint: float | Iterable[float],
    settling_band: float = 0.02,
    tail_fraction: float = 0.10,
) -> dict[str, float]:
    """Calculate common response metrics.

    Definitions used by this repository:
    - delay time: first crossing of 50% of commanded step.
    - rise time: t90 - t10.
    - peak time: time of extreme response in the commanded direction.
    - settling time: earliest time after which all remaining samples are inside
      ±settling_band of step magnitude (2% default).
    - maximum overshoot: excess beyond final SP relative to step magnitude.
    - steady-state error: final SP minus mean of last tail_fraction samples.

    Returns NaN when a metric cannot be defined from the provided data.
    """
    t = np.asarray(list(time_s), dtype=float)
    y = np.asarray(list(process_value), dtype=float)
    if np.isscalar(setpoint):
        sp = np.full_like(y, float(setpoint))
    else:
        sp = np.asarray(list(setpoint), dtype=float)

    if t.size < 3 or y.size != t.size or sp.size != t.size:
        raise ValueError("time, process_value, and setpoint must have equal length >= 3")

    mask = np.isfinite(t) & np.isfinite(y) & np.isfinite(sp)
    t, y, sp = t[mask], y[mask], sp[mask]
    if t.size < 3:
        raise ValueError("not enough finite samples")

    order = np.argsort(t)
    t, y, sp = t[order], y[order], sp[order]
    t = t - t[0]

    n_tail = max(3, int(round(t.size * tail_fraction)))
    y0 = float(np.mean(y[: min(3, y.size)]))
    spf = float(np.mean(sp[-n_tail:]))
    yss = float(np.mean(y[-n_tail:]))
    delta = spf - y0
    direction = 1.0 if delta >= 0 else -1.0
    amp = abs(delta)

    out = StepMetrics(initial_value=y0, final_setpoint=spf, final_value=yss)
    out.steady_state_error = spf - yss

    if amp <= 1e-12:
        return out.as_dict()

    l10 = y0 + 0.10 * delta
    l50 = y0 + 0.50 * delta
    l90 = y0 + 0.90 * delta

    t10 = _first_crossing(t, y, l10, direction)
    t50 = _first_crossing(t, y, l50, direction)
    t90 = _first_crossing(t, y, l90, direction)
    out.delay_time_s = t50
    if math.isfinite(t10) and math.isfinite(t90):
        out.rise_time_s = t90 - t10

    peak_idx = int(np.argmax(y) if direction > 0 else np.argmin(y))
    peak = float(y[peak_idx])
    out.peak_time_s = float(t[peak_idx])
    overshoot = direction * (peak - spf)
    out.maximum_overshoot_pct = max(0.0, overshoot / amp * 100.0)

    band = max(amp * settling_band, 1e-9)
    err = np.abs(y - spf)
    inside = err <= band
    settle = math.nan
    for i in range(inside.size):
        if bool(np.all(inside[i:])):
            settle = float(t[i])
            break
    out.settling_time_s = settle
    return out.as_dict()

File: shared/matlab/response_metrics.m
function M = response_metrics(t, y, sp)
%RESPONSE_METRICS Basic step-response metrics for measured lab data.
% t  : time [s]
% y  : process value
% sp : scalar or vector setpoint

t = t(:); y = y(:);
if isscalar(sp), sp = repmat(sp,size(y)); else, sp = sp(:); end
assert(numel(t)==numel(y) && numel(y)==numel(sp),'Input length mismatch');

ok = isfinite(t) & isfinite(y) & isfinite(sp);
t=t(ok); y=y(ok); sp=sp(ok);
t=t-t(1);

nTail=max(3,round(0.10*numel(y)));
y0=mean(y(1:min(3,end)));
spf=mean(sp(end-nTail+1:end));
yss=mean(y(end-nTail+1:end));
delta=spf-y0;
amp=abs(delta);

M.initial_value=y0;
M.final_setpoint=spf;
M.final_value=yss;
M.delay_time_s=NaN;
M.rise_time_s=NaN;
M.peak_time_s=NaN;
M.settling_time_s=NaN;
M.maximum_overshoot_pct=NaN;
M.steady_state_error=spf-yss;

if amp < eps, return; end
sgn=sign(delta); if sgn==0, sgn=1; end

cross = @(level) localCross(t,y,level,sgn);
t10=cross(y0+0.10*delta);
t50=cross(y0+0.50*delta);
t90=cross(y0+0.90*delta);
M.delay_time_s=t50;
if isfinite(t10) && isfinite(t90), M.rise_time_s=t90-t10; end

if sgn>0, [peak,idx]=max(y); else, [peak,idx]=min(y); end
M.peak_time_s=t(idx);
M.maximum_overshoot_pct=max(0,sgn*(peak-spf)/amp*100);

band=max(0.02*amp,eps);
inside=abs(y-spf)<=band;
for k=1:numel(inside)
    if all(inside(k:end))
        M.settling_time_s=t(k);
        break
    end
end
end

function tc = localCross(t,y,level,sgn)
if sgn>0, idx=find(y>=level,1,'first'); else, idx=find(y<=level,1,'first'); end
if isempty(idx), tc=NaN; else, tc=t(idx); end
end
